tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -763,6 +763,67 @@ class ToEtcExpr(
         val opsig = OperT1(Seq(a, BoolT1), BoolT1) // (a, Bool) => Bool
         mkExRefApp(opsig, Seq(sub, act))
 
+      // ******************************************** Variants **************************************************
+      case OperEx(VariantOper.variant, v @ ValEx(TlaStr(tagName)), valueEx) =>
+        val a = varPool.fresh
+        val b = varPool.fresh
+        // (Str, a) => T1a(a) | b
+        val opsig =
+          OperT1(Seq(StrT1, a), VariantT1(RowT1(b, tagName -> a)))
+        // For some reason, we required the field tag: Str to be present in the value type. We should revisit this.
+        mkExRefApp(opsig, Seq(v, valueEx))
+
+      case ex @ OperEx(VariantOper.variant, tag @ _, _) =>
+        throw new TypingInputException(s"The second argument of Variant must be a string, found: $tag", ex.ID)
+
+      case OperEx(VariantOper.`variantFilter`, v @ ValEx(TlaStr(tagName)), setEx) =>
+        val a = varPool.fresh
+        val b = varPool.fresh
+        // (Str, Set(T1a(a) | b)) => Set(a)
+        val opsig = OperT1(Seq(StrT1, SetT1(VariantT1(RowT1(b, tagName -> a)))), SetT1(a))
+        // For some reason, we required the field tag: Str to be present in the value type. We should revisit this.
+        mkExRefApp(opsig, Seq(v, setEx))
+
+      case ex @ OperEx(VariantOper.`variantFilter`, tag @ _, _) =>
+        throw new TypingInputException(s"The second argument of FilterByTag must be a string, found: $tag", ex.ID)
+
+      case OperEx(VariantOper.`variantMatch`, v @ ValEx(TlaStr(tagName)), variantEx, thenOper, elseOper) =>
+        val a = varPool.fresh
+        val b = varPool.fresh
+        val c = varPool.fresh
+        // a => c
+        val thenType = OperT1(Seq(a), c)
+        // Variant(b) => c
+        val elseType = OperT1(Seq(VariantT1(RowT1(b))), c)
+        // (Str, T1a(a) | b, thenOper, elseOper) => c
+        val operArgs =
+          Seq(
+              StrT1,
+              VariantT1(RowT1(b, tagName -> a)),
+              thenType,
+              elseType,
+          )
+
+        val opsig = OperT1(operArgs, c)
+        // For some reason, we required the field tag: Str to be present in the value type. We should revisit this.
+        mkExRefApp(opsig, Seq(v, variantEx, thenOper, elseOper))
+
+      case OperEx(VariantOper.`variantMatch`, tag @ _, _, _, _) =>
+        throw new TypingInputException(s"The second argument of MatchTag must be a string, found: $tag", ex.ID)
```

Comment:
Similar to Shon's previous comments.
```suggestion
        throw new TypingInputException(s"The first argument of MatchTag must be a string, found: $tag", ex.ID)
```